### PR TITLE
Create fip-draft_askv2.md

### PR DESCRIPTION
**Pull request by brendalee**

### CODE REVIEW

File: FIPS/fip-draft_askv2.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+discussions-to: <https://github.com/filecoin-project/FIPs/discussions/223>
+status: Draft
+type: <Technical>
+category (*only required for Standard Track): <Interface >
+created: <2021-12-13>
+spec-sections: 
+  - <section-id>
+  - <section-id>
+requires (*optional): <FIP number(s)>
+replaces (*optional): <FIP number(s)>
```
Comment (jennijuju) [inline FIPS/fip-draft_askv2.md:14]:
```suggestion
title: Storage Ask Protocol v2 Standard 
author: Brenda Lee (@brendalee), @Jiaying Wang (@jennijuju) , Why (@whyrusleeping), Jimmy Lee(@@jimmylee
jimmylee)
discussions-to: <https://github.com/filecoin-project/FIPs/discussions/225>
status: Draft
type: Technical
category (*only required for Standard Track): Interface 
created: <2021-12-13>
spec-sections: 
  - 2.7.1 (AskProtocol & Storage Ask)
```

File: FIPS/fip-draft_askv2.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+---
+
+<!--You can leave these HTML comments in your merged FIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new FIPs. Note that a FIP number will be assigned by an editor. When opening a pull request to submit your FIP, please use an abbreviated title in the filename, `fip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the FIP.-->
+The goal of Askv2 protocol is to help with awareness between the client and storage providers before a deal proposal. Before a client attempts to make a deal, they can send details of their proposed deal, and the storage provider can respond to the proposal (with accept or reject and other details such as price or earliest start epoch). Note that the ask protocol is not designed to handle back and forth deal negotiations.
```
Comment (jennijuju) [inline FIPS/fip-draft_askv2.md:21]:
```suggestion
The goal of AskProtocol v2 is to help with the [discovery](https://spec.filecoin.io/#section-systems.filecoin_markets.storage_market.discovery) between the client and storage providers before making a storage deal proposal. As the first step of the SP discovery, clients can send details of their proposed deal to the storage provider, and the storage provider should be able to respond to the proposal (with accept or reject, price and additional service offer). Note that the ask protocol is **not** designed to handle back and forth deal negotiations. 
```

File: FIPS/fip-draft_askv2.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+The goal of Askv2 protocol is to help with awareness between the client and storage providers before a deal proposal. Before a client attempts to make a deal, they can send details of their proposed deal, and the storage provider can respond to the proposal (with accept or reject and other details such as price or earliest start epoch). Note that the ask protocol is not designed to handle back and forth deal negotiations. 
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+Askv2 aims to improve the existing ask protocol to help the client understand whether or not a storage provider is likely to accept their deal. 
+
+The proposed flow is simple. The client sends an ask request with information about the storage deal they would like to make. The ask response by the storage provider indicates whether or not the deal will be accepted with specifics on conditions (for example, how long the offer is valid for, and what the price to store is). More details on information the client/storage provider will include in the messages can be found in the specification section.
```
Comment (jennijuju) [inline FIPS/fip-draft_askv2.md:27]:
```suggestion
AskProtocol v2 aims to improve the existing AskProtocol to help the client understand whether or not a storage provider is likely to accept their deal before they send a signed proposal. 

The proposed flow is simple. The client sends an ask request with information about the storage deal they would like to make. The ask response by the storage provider indicates whether or not the deal will be accepted with specifics on conditions (for example, how long the offer is valid for, and what the price to store is). More details about the AskProtocol v2 request/response can be found in the specification section. 
```

File: FIPS/fip-draft_askv2.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  - Need clearly defined rejection codes
+  - Start with HTTP codes and build from there
+  - Option for storage provider to specify additional rejection codes
+* Price per GB per epoch
+  - Single price field
+* How long the offer is valid for
+  - Expiry time
+  - Not binding, just for client convenience
+* Start epoch of deal
+* Message freeform text
```
Comment (jennijuju) [inline FIPS/fip-draft_askv2.md:64]:
```suggestion
### Client Request of AskProtocol v2

For the AskProtocol v2 request, the client provides the following information: 

```go=
type StorageAskRequest struct {
  ClientAddress Address
  PeerId  *abi.PeerID 
  PieceSize abi.PaddedPieceSize
  Duration int64
  StartEpoch abi.ChainEpoch 
  Verified bool
  TransferProtocol string
  Message string
```
where:
* `ClientAddress`: is the wallet/account address that the client will be used for signing the deal proposal and making/paying the storage deal
* `PeerId`: is the peerID of the client's node, this is an optional field
* `PieceSize`: is the padded size of the piece in the deal
* `Duration`: is the deal duration 
* `StartEpoch`: is the start epoch of the deal
* `Verified`: if the deal is a verified deal 
* `Transfer protocol`: is the protocol that will be used for data transfer, for example, graphsync, physical. Aside from a list of defined protocols <TODO>, the client/storage provider may have the option to specify/add additional transfer protocols 
* `Message`: is a free-form field set by the client to the storage provider
*Signature*: the request should be signed by the client address  

### Storage Provider Response of AskProtocol v2

For the response, the provider will respond with the following information: 

```go=
type StorageAskResponse struct {
  Query StorageAskRequest
  WillAccept bool
  RejectionCode uint
  Price abi.TokenAmount
  Expiry time.Time
  DealStartEpoch abi.ChainEpoch
  Message string
}
```
where:
* `Query`: is the `StorageAskRequest` that was sent by the client
* `WillAccept`: is whether the storage provider will accept or reject the deal
* `RejectionCode`: is codes for rejection reasons
  - Need clearly defined rejection codes
  - Start with HTTP codes and build from there
  - Option for storage provider to specify additional rejection codes
* `Price`: is the offering price per GB per epoch.
* `Expiry`: is how long the offer is valid for, this is not a piece of binding information but just for client's insights.
* `DealStartEpoch`: is how the storage provider is offering to the deal fully committed to the network within a sector
* `Message: is a free-form field set by the storage provider to the client
```

**Review by anorth — COMMENTED** [review]
Following the process discussion in https://github.com/filecoin-project/FIPs/discussions/215, I do not think this proposal is sufficiently fleshed out to warrant a PR yet. Much more should be worked out in the discussion forum or off-Github drafts first.

cc @kaitlin-beegle

File: FIPS/fip-draft_askv2.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  - Option for client to specify additional transfer protocols
+  - Will have a set of defined protocols, but allow SPs to add more
+* Message freeform text 
+* Signature by client 
+
+For the response, the provider will respond with the following information: 
+* Will accept or reject the deal
+* Rejection code
+  - Need clearly defined rejection codes
+  - Start with HTTP codes and build from there
```
[the remark this answers is not in the document]
Comment (RobQuistNL) [reply FIPS/fip-draft_askv2.md]: We should keep HTTP 418.

File: FIPS/fip-draft_askv2.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+Askv2 aims to improve the existing ask protocol to help the client understand whether or not a storage provider is likely to accept their deal. 
+
+The proposed flow is simple. The client sends an ask request with information about the storage deal they would like to make. The ask response by the storage provider indicates whether or not the deal will be accepted with specifics on conditions (for example, how long the offer is valid for, and what the price to store is). More details on information the client/storage provider will include in the messages can be found in the specification section. 
+
+
+## Change Motivation
+<!--The motivation is critical for FIPs that want to change the Filecoin protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the FIP solves. FIP submissions without sufficient motivation may be rejected outright.-->
+Clients need to be able to share details of the storage deal they want to make. Furthermore, they need to know what capabilities and features a Storage Provider supports, so as to understand what Storage Providers they can successfully make a deal with.
```
[the remark this answers is not in the document]
Comment (brendalee) [reply FIPS/fip-draft_askv2.md:29]: Thanks Alex! With the existing ask protocol, the client does not have a method to share information about their storage deal; rather, only the storage provider can share precondition parameters by which they will choose to accept or reject a deal. [Existing StorageAsk](https://github.com/filecoin-project/go-fil-markets/blob/e111ec29d24d02586718dbab465571e0b8220a24/storagemarket/types.go#L48)

File: FIPS/fip-draft_askv2.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+* Rejection code
+  - Need clearly defined rejection codes
+  - Start with HTTP codes and build from there
+  - Option for storage provider to specify additional rejection codes
+* Price per GB per epoch
+  - Single price field
+* How long the offer is valid for
+  - Expiry time
+  - Not binding, just for client convenience
+* Start epoch of deal
```
[the remark this answers is not in the document]
Comment (brendalee) [reply FIPS/fip-draft_askv2.md]: Yes, that's correct. An example use case would be if clients want to make multiple deals with the same storage provider, then they would know what timing would be convenient to do dealmaking.

File: FIPS/fip-draft_askv2.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (brendalee) [reply FIPS/fip-draft_askv2.md]: @anorth - do you have additional suggestions for what error/rejection codes should be as an alternative to HTTP codes? Would be great to get your input/feedback in the discussion.

**Comment by kaitlin-beegle** [comment]
Closing this PR since an updated version (as an FRC) was just pushed and merged.

**Comment by brendalee** [comment]
@kaitlin-beegle this is not yet covered in boost, but hopefully can be worked on in the near future

**Pull request closed without merging**